run-make `-Zbuild-std` tests fail on `compiler_builtins` on older branches

I tried: merging backports on 1.82-beta (#130569)

I expected to see this happen: Test successful

Instead, this happened: The job `dist-various-1` failed!

```
failures:
    [run-make] tests/run-make/compiler-builtins
    [run-make] tests/run-make/thumb-none-cortex-m
```

Both tests use `-Zbuild-std` and failed to build `compiler_builtins v0.1.126`:

```
error[E0433]: failed to resolve: could not find `naked_asm` in `arch`
##[error]  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.126/src/arm.rs:26:21
   |
26 |         core::arch::naked_asm!(
   |                     ^^^^^^^^^ could not find `naked_asm` in `arch`

error[E0433]: failed to resolve: could not find `naked_asm` in `arch`
##[error]  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.126/src/arm.rs:39:21
   |
39 |         core::arch::naked_asm!(
   |                     ^^^^^^^^^ could not find `naked_asm` in `arch`

error[E0433]: failed to resolve: could not find `naked_asm` in `arch`
##[error]  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.126/src/arm.rs:54:21
   |
54 |         core::arch::naked_asm!(
   |                     ^^^^^^^^^ could not find `naked_asm` in `arch`

error[E0433]: failed to resolve: could not find `naked_asm` in `arch`
##[error]  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.126/src/arm.rs:66:21
   |
66 |         core::arch::naked_asm!(
   |                     ^^^^^^^^^ could not find `naked_asm` in `arch`
```

Note that beta's `library/Cargo.lock` only has `compiler_builtins v0.1.123`, so it appears cargo isn't heeding that with `-Zbuild-std`.

`arch::naked_asm!` was added in #130146 and started being used in rust-lang/compiler-builtins#686 for 0.1.126.

To unblock beta, I'm going to add `//@ only-nightly` to those tests, as suggested on my PR, especially since `-Zbuild-std` is unstable anyway. This particular failure shouldn't be an issue when the current nightlies are promoted to the next beta, but we should think about this kind of failure mode in the long term too.

--- a/src/bootstrap/src/core/build_steps/test.rs
+++ b/src/bootstrap/src/core/build_steps/test.rs
@@ -1733,6 +1733,11 @@ NOTE: if you're sure you want to do this, please open an issue as to why. In the
 
         let is_rustdoc = suite.ends_with("rustdoc-ui") || suite.ends_with("rustdoc-js");
 
+        if mode == "run-make" {
+            let cargo = builder.ensure(tool::Cargo { compiler, target: compiler.host });
+            cmd.arg("--cargo-path").arg(cargo);
+        }
+
         // Avoid depending on rustdoc when we don't need it.
         if mode == "rustdoc"
             || mode == "run-make"

--- a/src/tools/compiletest/src/common.rs
+++ b/src/tools/compiletest/src/common.rs
@@ -183,6 +183,9 @@ pub struct Config {
     /// The rustc executable.
     pub rustc_path: PathBuf,
 
+    /// The cargo executable.
+    pub cargo_path: Option<PathBuf>,
+
     /// The rustdoc executable.
     pub rustdoc_path: Option<PathBuf>,
 

--- a/src/tools/compiletest/src/lib.rs
+++ b/src/tools/compiletest/src/lib.rs
@@ -47,6 +47,7 @@ pub fn parse_config(args: Vec<String>) -> Config {
     opts.reqopt("", "compile-lib-path", "path to host shared libraries", "PATH")
         .reqopt("", "run-lib-path", "path to target shared libraries", "PATH")
         .reqopt("", "rustc-path", "path to rustc to use for compiling", "PATH")
+        .optopt("", "cargo-path", "path to cargo to use for compiling", "PATH")
         .optopt("", "rustdoc-path", "path to rustdoc to use for compiling", "PATH")
         .optopt("", "coverage-dump-path", "path to coverage-dump to use in tests", "PATH")
         .reqopt("", "python", "path to python to use for doc tests", "PATH")
@@ -260,6 +261,7 @@ pub fn parse_config(args: Vec<String>) -> Config {
         compile_lib_path: make_absolute(opt_path(matches, "compile-lib-path")),
         run_lib_path: make_absolute(opt_path(matches, "run-lib-path")),
         rustc_path: opt_path(matches, "rustc-path"),
+        cargo_path: matches.opt_str("cargo-path").map(PathBuf::from),
         rustdoc_path: matches.opt_str("rustdoc-path").map(PathBuf::from),
         coverage_dump_path: matches.opt_str("coverage-dump-path").map(PathBuf::from),
         python: matches.opt_str("python").unwrap(),
@@ -364,6 +366,7 @@ pub fn log_config(config: &Config) {
     logv(c, format!("compile_lib_path: {:?}", config.compile_lib_path));
     logv(c, format!("run_lib_path: {:?}", config.run_lib_path));
     logv(c, format!("rustc_path: {:?}", config.rustc_path.display()));
+    logv(c, format!("cargo_path: {:?}", config.cargo_path));
     logv(c, format!("rustdoc_path: {:?}", config.rustdoc_path));
     logv(c, format!("src_base: {:?}", config.src_base.display()));
     logv(c, format!("build_base: {:?}", config.build_base.display()));

--- a/src/tools/compiletest/src/runtest/run_make.rs
+++ b/src/tools/compiletest/src/runtest/run_make.rs
@@ -61,6 +61,10 @@ impl TestCx<'_> {
             .env_remove("MFLAGS")
             .env_remove("CARGO_MAKEFLAGS");
 
+        if let Some(ref cargo) = self.config.cargo_path {
+            cmd.env("CARGO", cwd.join(cargo));
+        }
+
         if let Some(ref rustdoc) = self.config.rustdoc_path {
             cmd.env("RUSTDOC", cwd.join(rustdoc));
         }
@@ -408,6 +412,10 @@ impl TestCx<'_> {
             // Provide which LLVM components are available (e.g. which LLVM components are provided
             // through a specific CI runner).
             .env("LLVM_COMPONENTS", &self.config.llvm_components);
+
+        if let Some(ref cargo) = self.config.cargo_path {
+            cmd.env("CARGO", source_root.join(cargo));
+        }
 
         if let Some(ref rustdoc) = self.config.rustdoc_path {
             cmd.env("RUSTDOC", source_root.join(rustdoc));

--- a/tests/run-make/compiler-builtins/rmake.rs
+++ b/tests/run-make/compiler-builtins/rmake.rs
@@ -33,8 +33,8 @@ fn main() {
 
     let path = env_var("PATH");
     let rustc = env_var("RUSTC");
-    let bootstrap_cargo = env_var("BOOTSTRAP_CARGO");
-    let mut cmd = cmd(bootstrap_cargo);
+    let cargo = env_var("CARGO");
+    let mut cmd = cmd(cargo);
     cmd.args(&[
         "build",
         "--manifest-path",

--- a/tests/run-make/thumb-none-cortex-m/rmake.rs
+++ b/tests/run-make/thumb-none-cortex-m/rmake.rs
@@ -36,10 +36,10 @@ fn main() {
 
     let path = env_var("PATH");
     let rustc = env_var("RUSTC");
-    let bootstrap_cargo = env_var("BOOTSTRAP_CARGO");
-    // FIXME: extract bootstrap cargo invocations to a proper command
+    let cargo = env_var("CARGO");
+    // FIXME: extract cargo invocations to a proper command
     // https://github.com/rust-lang/rust/issues/128734
-    let mut cmd = cmd(bootstrap_cargo);
+    let mut cmd = cmd(cargo);
     cmd.args(&[
         "build",
         "--manifest-path",

